File: backend/app/schemas/election.py
from typing import List, Optional
from pydantic import BaseModel
from datetime import datetime

# Candidate Schemas
class CandidateBase(BaseModel):
    name: str
    bio: Optional[str] = None
    photo_url: Optional[str] = None

class CandidateCreate(CandidateBase):
    pass

class Candidate(CandidateBase):
    id: int
    election_id: int

    class Config:
        from_attributes = True

# Vote Schemas
class VoteBase(BaseModel):
    candidate_id: int

class VoteCreate(VoteBase):
    pass

class Vote(VoteBase):
    id: int
    election_id: int
    user_id: int
    timestamp: datetime

    class Config:
        from_attributes = True

# Election Schemas
class ElectionBase(BaseModel):
    title: str
    description: Optional[str] = None
    start_date: Optional[datetime] = None
    end_date: datetime
    is_active: bool = True

class ElectionCreate(ElectionBase):
    pass

class ElectionUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    is_active: Optional[bool] = None

class Election(ElectionBase):
    id: int
    candidates: List[Candidate] = []

    class Config:
        from_attributes = True

class ElectionResults(BaseModel):
    election_id: int
    title: str
    results: List[dict] # { "candidate_id": int, "name": str, "votes": int }

File: frontend/src/services/election.ts
import { fetchJson } from "@/lib/api";

export interface Candidate {
    id: number;
    election_id: number;
    name: string;
    bio?: string;
    photo_url?: string;
}

export interface Election {
    id: number;
    title: string;
    description?: string;
    start_date: string;
    end_date: string;
    is_active: boolean;
    candidates: Candidate[];
}

export interface ElectionCreate {
    title: string;
    description: string;
    end_date: string;
    is_active?: boolean;
    position_id?: number | null;
}

export interface CandidateCreate {
    name: string;
    bio?: string;
    photo_url?: string;
}

export interface ElectionResults {
    election_id: number;
    title: string;
    results: {
        candidate_id: number;
        name: string;
        votes: number;
    }[];
}

export const electionService = {
    async getElections(): Promise<Election[]> {
        return fetchJson<Election[]>("/api/v1/elections/");
    },

    async createElection(election: ElectionCreate): Promise<Election> {
        return fetchJson<Election>("/api/v1/elections/", {
            method: "POST",
            body: JSON.stringify(election),
        });
    },

    async addCandidate(electionId: number, candidate: CandidateCreate): Promise<Candidate> {
        return fetchJson<Candidate>(`/api/v1/elections/${electionId}/candidates`, {
            method: "POST",
            body: JSON.stringify(candidate),
        });
    },

    async castVote(electionId: number, candidateId: number): Promise<any> {
        return fetchJson(`/api/v1/elections/${electionId}/vote`, {
            method: "POST",
            body: JSON.stringify({ candidate_id: candidateId }),
        });
    },

    async getResults(electionId: number): Promise<ElectionResults> {
        return fetchJson<ElectionResults>(`/api/v1/elections/${electionId}/results`);
    }
};
